code/game/machinery/ecto_sniffer.dm

```diff
@@ -0,0 +1,79 @@
+/obj/machinery/ecto_sniffer
+	name = "ectoscopic sniffer"
+	desc = "A highly sensitive parascientific instrument calibrated to detect the slightest whiff of ectoplasm."
+	icon = 'icons/obj/machines/research.dmi'
+	icon_state = "ecto_sniffer"
+	density = FALSE
+	anchored = FALSE
+	pass_flags = PASSTABLE
+	///determines if the device if the power switch is turned on or off. Useful if the ghosts are too annoying.
+	var/on = TRUE
+	///If this var set to false the ghosts will not be able interact with the machine, say if the machine is silently disabled by cutting the internal wire.
+	var/sensor_enabled = TRUE
+	///List of weakref datums containing mobs who have recently activated the device, mobs on this list are prohibited from activating the device untill their residue decays.
+	var/list/ectoplasmic_residues = list()
+
+/obj/machinery/ecto_sniffer/Initialize()
+	. = ..()
+	wires = new/datum/wires/ecto_sniffer(src)
+
+/obj/machinery/ecto_sniffer/attack_ghost(mob/user)
+	. = ..()
+	if(!is_operational || !on || !sensor_enabled)
+		return
+
+	for(var/datum/weakref/spirit_residue in ectoplasmic_residues)
+		if(IS_WEAKREF_OF(user, spirit_residue))
+			return ..()
+	activate(user)
+
+/obj/machinery/ecto_sniffer/proc/activate(mob/activator)
+	flick("ecto_sniffer_flick", src)
+	playsound(loc, 'sound/machines/ping.ogg', 20)
+	use_power(10)
+	if(activator)
+		ectoplasmic_residues += WEAKREF(activator)
+		addtimer(CALLBACK(src, .proc/clear_residue, activator), 15 SECONDS)
```

Comment:
Can't we also have the argument passed to the timer system a weakrefence to the ghost as well?

Reply:
done